node/src/main/scala/io/casperlabs/node/configuration/ChainSpec.scala

```
@@ -123,76 +128,132 @@ object ChainSpec extends ParserImplicits {
+/** Resolve to normal files. */
+object FileResolver extends Resolver {
+  override def asBytes(path: Path) =
+    Utils.readBytes(path)
 
-  private def withManifest[A, B](dir: Path, parseManifest: (=> Source) => ValidatedNel[String, A])(
-      read: A => Either[String, B]
-  ): ValidatedNel[String, B] = {
-    val manifest = new File(dir.toFile, "manifest.toml")
-    if (!manifest.exists)
-      Validated.invalidNel(s"Manifest file '$manifest' is missing!")
-    else {
-      parseManifest(Source.fromFile(manifest)) andThen { conf =>
-        read(conf)
-          .leftMap(err => s"Could not read chainspec sub-directory $dir: $err")
-          .toValidatedNel
-      }
+  override def asSource(path: Path) =
+    if (!path.toFile.exists)
+      s"File '$path' is missing!".asLeft[Source]
+    else
+      Source.fromFile(path.toFile).asRight[String]
+
+  override def asString(path: Path) =
+    Utils.readFile(path)
+
+  override def listFiles(path: Path) =
+    path.toFile.listFiles.sortBy(_.getName).map(_.toPath).toList
+}
+
+/** Resolve paths in resources, unless an override in the data directory exists. */
+class ResourceResolver(dataDir: Path) extends Resolver {
+  override def asBytes(path: Path) =
+    read(path, readResourceBytes, FileResolver.asBytes)
+
+  override def asSource(path: Path) =
+    read(path, x => Source.fromResource(x.toString), FileResolver.asSource)
+
+  override def asString(path: Path) =
+    asSource(path).flatMap(src => Utils.readFile(src))
+
+  override def listFiles(path: Path) =
+    ResourceResolver.listFilesInResources(path)
+
+  private def read[T](
+      path: Path,
+      fromResource: Path => T,
+      fromFile: Path => Either[String, T]
+  ): Either[String, T] =
+    if (path.isAbsolute) {
+      fromFile(path)
+    } else {
+      val over = dataDir.resolve(path).toFile
+      if (over.exists) fromFile(over.toPath)
+      else
+        Try(fromResource(path)).fold(
```

So if `path` is not absolute AND `dataDir/path` doesn't exist either then it will try to read only from `path`? Why not return an error that `dataDir/path` is not valid?

---

It will try to read from `resources/path`, i.e. from the JAR. `dataDir/path` is your way to shadow a file in `resources`. For example to supplement a mint or PoS installer, or use your own account.csv, just drop them at the right place. It just avoids the need for extracting anything up front explicitly, but leaves it open for you to opt into modifying bits if you know they exist.